firedrake/preconditioners/fdm.py

```diff
@@ -0,0 +1,1036 @@
+from functools import lru_cache, partial
+
+from pyop2 import op2
+from pyop2.sparsity import get_preallocation
+
+from ufl import FiniteElement, VectorElement, TensorElement
+from ufl import FacetNormal, Jacobian, JacobianDeterminant, JacobianInverse
+from ufl import as_tensor, diag_vector, dot, dx, indices, inner, inv
+from ufl.algorithms.ad import expand_derivatives
+
+from firedrake.petsc import PETSc
+from firedrake.preconditioners.base import PCBase
+from firedrake.preconditioners.patch import bcdofs
+from firedrake.preconditioners.pmg import get_permuted_map, tensor_product_space_query
+from firedrake.utils import IntType_c
+from firedrake.dmhooks import get_function_space, get_appctx
+import firedrake.dmhooks as dmhooks
+import firedrake
+import numpy
+
+try:
+    from scipy.linalg import eigh
+
+    def sym_eig(A, B):
+        return eigh(A, B)
+except ImportError:
+    import numpy.linalg as npla
+
+    def sym_eig(A, B):
+        L = npla.cholesky(B)
+        Linv = npla.inv(L)
+        C = numpy.dot(Linv, numpy.dot(A, Linv.T))
+        Z, W = npla.eigh(C)
+        V = numpy.dot(Linv.T, W)
+        return Z, V
+
+
+class FDMPC(PCBase):
+
+    _prefix = "fdm_"
+
+    def initialize(self, pc):
+        A, P = pc.getOperators()
+
+        # Read options
+        prefix = pc.getOptionsPrefix()
+        options_prefix = prefix + self._prefix
+        # opts = PETSc.Options(options_prefix)
+
+        dm = pc.getDM()
+        V = get_function_space(dm)
+        use_tensorproduct, N, ndim, family, _ = tensor_product_space_query(V)
+
+        if not use_tensorproduct or (family <= {"RTCE", "NCE"}):
+            raise ValueError("The element %s is not supported by FDMPC." % (V.ufl_element(), ))
+
+        needs_interior_facet = not (family <= {"Q", "Lagrange"})
+        Nq = 2*N+1  # quadrature degree
+
+        self.mesh = V.mesh()
+        self.uf = firedrake.Function(V)
+        self.uc = firedrake.Function(V)
+        self.cell_node_map = get_permuted_map(V)
+
+        # Get problem solution and bcs
+        solverctx = get_appctx(dm)
+        self.u = solverctx._problem.u
+        self.bcs = solverctx.bcs_F
+
+        if len(self.bcs) > 0:
+            self.bc_nodes = numpy.unique(numpy.concatenate([bcdofs(bc, ghost=False)
+                                                            for bc in self.bcs]))
+        else:
+            self.bc_nodes = numpy.empty(0, dtype=PETSc.IntType)
+
+        bcflags = self.get_bc_flags(V, self.mesh, self.bcs, solverctx._problem.J)
+
+        self.weight = self.multiplicity(V)
+        with self.weight.dat.vec as w:
+            w.reciprocal()
+
+        # Get form coefficients
+        appctx = self.get_appctx(pc)
+        eta = appctx.get("eta", (N+1)*(N+ndim))  # interior penalty parameter
```

Comment:
Can you document (perhaps in the docstring for FDMPC) what the "viscosity" and "reaction" terms are? How can a general reader know whether this applies to their problem or not?